### PR TITLE
Improved Radial Basis Function Interpolation

### PR DESCRIPTION
**Pull request by Jwink3101**
#### Reference issue
This is a first attempt at implimented what I sketched in #9904 and also addresses #7157.

#### What does this implement/fix?
`scipy.interpolate.RadialBasisFunction` as an eventual replacement for `scipy.interpolate.Rbf`

#### Additional information

Things I think need to be addressed are in bullets

I also added an np.deprecate to the old Rbf though it should probably be updated with concrete deprecation plans

* This needs to be checked to make sure it is in the right format

I added a few tests but there are likely more worth considering. I am not sure the best approach to test things like how well it does away from the training data without hard coding results. That is a qualitative measure and will depend the settings. I can tell you in practice it works very well if your data is smooth

* I am not sure if this test is being run in the test suite
* I would like some ideas on how to best assess quality of the interp. It may even be out of scope!

The documentation inline is very detailed. May be too detailed? It also includes references. [1] and [2] are the main ones but [3] and [4] are referenced in the documentation for resources on how to choose epsilon

* Does the documentation need to be mirrored elsehwere?

Some things I considered but did *not* impliment:

* Leave-one-out error estimate (see discussions in #9904)
* Scaling the input to a unit-hypercube. I note it as a tip, but it is not done. It would be super easy to add. Thoughts?

Anything else? This is my first major contribution so I may be missing some key steps.

Thanks!

### CODE REVIEW

**Comment by Jwink3101** [comment]
I suspect the test failures are from my deprecating `Rbf`. Can someone with more experience confirm? And if that is the case, where to do I update them to prevent that?

Any help would be appreciated!

**Comment by pv** [comment]
Let's do the deprecation in a separate PR --- we'll e.g. need to decide whether RBF would be going to be removed eventually.

**Comment by Jwink3101** [comment]
@pv Sounds good. I updated it

(I hope it is okay that I merged upstream into my branch to make sure there were no issues. I didn't rebase it since I wasn't sure what SciPy's rebase policy is and whether or not to rebase is a religious discussion!)

File: scipy/interpolate/radial_basis_interp.py [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+    which is then solved. Given c and g, (2) can be used to solve for
+    any given x
+    
+    References:
+    ----------  
+    [1] G. B. Wright. Radial Basis Function Interpolation: Numerical and 
+        Analytical Developments. PhD thesis, University of Colorado, 2003.
+
+    [2] SciPy version 1.1.0
+        https://github.com/scipy/scipy/blob/v1.1.0/scipy/interpolate/rbf.py
```
Comment (ev-br) [inline scipy/interpolate/radial_basis_interp.py]: I don't think this reference is worth including in the scipy docs itself :-).

File: scipy/interpolate/tests/test_radial_basis.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    
+    for kernel,dim,degree in itertools.product(kernels,dims,degrees):
+        if dim is None:
+            x = X[:,0] # shape N,
+            f = np.sin(2*np.pi*x)
+        else:
+            x = X[:,:dim]
+            f = np.sin(2*np.pi*x).sum(axis=1)
+        
+        epsilon = None # Use the default
```
Comment (ev-br) [inline scipy/interpolate/tests/test_radial_basis.py]: there should be a test with an explicit epsilon, too

**Review by ev-br — COMMENTED** [review]
Looks like a great start.

I only had a cursory look and left a few mostly cosmetic comments (a thorough review will have to wait for some other day, sadly). Bigger questions:

- it definitely needs more tests
- as a start, I'd go through the existing `rbf` tests. 
- maybe it's worth thinking a bit about the API. Maybe there's a way of encapsulating the epsilon-related logic into kernels? So that there's no epsilon unless a kernel needs it. Or maybe the status quo is OK.
- Not sure what the `smooth` parameter does.

At any rate a new feature should go through the mailing list.

**Comment by Jwink3101** [comment]
That's a lot of very constructive comments. I will try to tackle them over the holidays. Thanks for the feedback.

**Comment by rgommers** [comment]
This has been quiet for 1.5 years, and is now superceded by the new RBFInterpolator we just merged in gh-13595. I think it ticks all the boxes for wishes people had for `Rbf`. So I'll close this PR. Thanks for your contribution @Jwink3101! If you'd want to try out `RBFInterpolator` and see if it does all you need, that'd be great.

**Pull request closed without merging**